File: ios/App/UniMusicSyncPlugin.swift
import Capacitor
import UniformTypeIdentifiers
import UniMusicSync

@objc(UniMusicSyncPlugin)
public class UniMusicSyncPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "UniMusicSyncPlugin"
    public let jsName = "UniMusicSync"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "createNamespace", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deleteNamespace", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAuthor", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getNodeId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getFiles", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "writeFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deleteFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "readFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "readFileHash", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "share", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "export", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "exportHash", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "import", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sync", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reconnect", returnType: CAPPluginReturnPromise),
    ]

    public private(set) var uniMusicSync: UniMusicSync? = .none

    deinit {
        if let uniMusicSync {
            Task {
                try? await uniMusicSync.shutdown()
            }
        }
    }

    override public func load() {
        Task {
            let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
            let syncPath = documentsPath.appendingPathComponent("UniMusicSync")
            do {
                uniMusicSync = try await UniMusicSync(syncPath.path)
                print("Initialized UniMusicSync")
            } catch {
                print("Failed to initialize UniMusicSync: \(error)")
            }
        }
    }

    @objc public func createNamespace(_ call: CAPPluginCall) {
        guard let uniMusicSync else {
            call.reject("UniMusicSync is not initialized")
            return
        }

        Task {
            do {
                let namespace = try await uniMusicSync.createNamespace()
                call.resolve([
                    "namespace": namespace,
                ])
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func deleteNamespace(_ call: CAPPluginCall) {
        guard let uniMusicSync else {
            call.reject("UniMusicSync is not initialized")
            return
        }

        guard let namespace = call.getString("namespace") else {
            call.reject("You must provide a 'namespace' option")
            return
        }

        Task {
            do {
                try await uniMusicSync.deleteNamespace(namespace: namespace)
                call.resolve()
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func getAuthor(_ call: CAPPluginCall) {
        guard let uniMusicSync else {
            call.reject("UniMusicSync is not initialized")
            return
        }

        Task {
            do {
                let author = try await uniMusicSync.getAuthor()
                call.resolve([
                    "author": author,
                ])
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func getNodeId(_ call: CAPPluginCall) {
        guard let uniMusicSync else {
            call.reject("UniMusicSync is not initialized")
            return
        }

        Task {
            let nodeId = await uniMusicSync.getNodeId()
            call.resolve([
                "nodeId": nodeId,
            ])
        }
    }

    @objc public func getFiles(_ call: CAPPluginCall) {
        guard let uniMusicSync else {
            call.reject("UniMusicSync is not initialized")
            return
        }

        guard let namespace = call.getString("namespace") else {
            call.reject("You must provide a 'namespace' option")
            return
        }

        Task {
            do {
                let files = try await uniMusicSync.getFiles(namespace)

                var filesObject: PluginCallResultData = [:]
                for entry in files {
                    if entry.isEmpty() {
                        continue
                    }

                    let key = entry.key()
                    filesObject[key] = [
                        "key": key,
                        "author": entry.author(),
                        "timestamp": entry.timestamp(),
                        "contentHash": entry.contentHash(),
                        "contentLen": entry.contentLen(),
                    ]
                }

                call.resolve(filesObject)
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func writeFile(_ call: CAPPluginCall) {
        guard let uniMusicSync else {
            call.reject("UniMusicSync is not initialized")
            return
        }

        guard let namespace = call.getString("namespace") else {
            call.reject("You must provide a 'namespace' option")
            return
        }

        guard let syncPath = call.getString("syncPath") else {
            call.reject("You must provide a 'syncPath' option")
            return
        }

        guard let sourcePath = call.getString("sourcePath"),
              let sourcePathUrl = URL(string: sourcePath)
        else {
            call.reject("You must provide a valid file url for 'sourcePath' option")
            return
        }

        Task {
            do {
                let data = try Data(contentsOf: sourcePathUrl)
                let fileHash = try await uniMusicSync.writeFile(
                    namespace,
                    syncPath,
                    data
                )
                call.resolve([
                    "fileHash": fileHash,
                ])
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func deleteFile(_ call: CAPPluginCall) {
        guard let uniMusicSync else {
            call.reject("UniMusicSync is not initialized")
            return
        }

        guard let namespace = call.getString("namespace") else {
            call.reject("You must provide a 'namespace' option")
            return
        }

        guard let syncPath = call.getString("syncPath") else {
            call.reject("You must provide a 'syncPath' option")
            return
        }

        Task {
            do {
                let tombstoneHash = try await uniMusicSync.deleteFile(namespace, syncPath)
                call.resolve()
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func readFile(_ call: CAPPluginCall) {
        guard let uniMusicSync else {
            call.reject("UniMusicSync is not initialized")
            return
        }

        guard let namespace = call.getString("namespace") else {
            call.reject("You must provide a 'namespace' option")
            return
        }

        guard let syncPath = call.getString("syncPath") else {
            call.reject("You must provide a 'syncPath' option")
            return
        }

        Task {
            do {
                let data = try await uniMusicSync.readFile(namespace, syncPath)
                let dataType = UTType(filenameExtension: (syncPath as NSString).pathExtension)
                let server = HTTPResponseServer(body: data, type: dataType)
                let url = try await server.start()
                call.resolve([
                    "url": url.absoluteString,
                ])
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func readFileHash(_ call: CAPPluginCall) {
        guard let uniMusicSync else {
            call.reject("UniMusicSync is not initialized")
            return
        }

        guard let fileHash = call.getString("fileHash") else {
            call.reject("You must provide a 'fileHash' option")
            return
        }

        Task {
            do {
                let data = try await uniMusicSync.readFileHash(fileHash)
                let server = HTTPResponseServer(body: data, type: .data)
                let url = try await server.start()
                call.resolve([
                    "url": url,
                ])
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func export(_ call: CAPPluginCall) {
        guard let uniMusicSync else {
            call.reject("UniMusicSync is not initialized")
            return
        }

        guard let namespace = call.getString("namespace") else {
            call.reject("You must provide a 'namespace' option")
            return
        }

        guard let syncPath = call.getString("syncPath") else {
            call.reject("You must provide a 'syncPath' option")
            return
        }

        guard let destinationPath = call.getString("destinationPath"),
              let _ = URL(string: destinationPath)
        else {
            call.reject("You must provide a valid file url for 'destinationPath' option")
            return
        }

        Task {
            do {
                try await uniMusicSync.export(namespace, syncPath, destinationPath)
                call.resolve()
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func exportHash(_ call: CAPPluginCall) {
        guard let uniMusicSync else {
            call.reject("UniMusicSync is not initialized")
            return
        }

        guard let fileHash = call.getString("fileHash") else {
            call.reject("You must provide a 'fileHash' option")
            return
        }

        guard let destinationPath = call.getString("destinationPath"),
              let _ = URL(string: destinationPath)
        else {
            call.reject("You must provide a valid file url for 'destination' option")
            return
        }

        Task {
            do {
                try await uniMusicSync.exportHash(fileHash, destinationPath)
                call.resolve()
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func share(_ call: CAPPluginCall) {
        guard let uniMusicSync else {
            call.reject("UniMusicSync is not initialized")
            return
        }

        guard let namespace = call.getString("namespace") else {
            call.reject("You must provide a 'namespace' option")
            return
        }

        Task {
            do {
                let ticket = try await uniMusicSync.share(namespace)
                call.resolve([
                    "ticket": ticket,
                ])
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func `import`(_ call: CAPPluginCall) {
        guard let uniMusicSync else {
            call.reject("UniMusicSync is not initialized")
            return
        }

        guard let ticket = call.getString("ticket") else {
            call.reject("You must provide a 'namespaceId' option")
            return
        }

        Task {
            do {
                let namespace = try await uniMusicSync.import(ticket)
                call.resolve([
                    "namespace": namespace,
                ])
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func sync(_ call: CAPPluginCall) {
        guard let uniMusicSync else {
            call.reject("UniMusicSync is not initialized")
            return
        }

        guard let namespace = call.getString("namespace") else {
            call.reject("You must provide a 'namespace' option")
            return
        }

        Task {
            do {
                try await uniMusicSync.sync(namespace)
                call.resolve()
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func reconnect(_ call: CAPPluginCall) {
        guard let uniMusicSync else {
            call.reject("UniMusicSync is not initialized")
            return
        }

        Task {
            await uniMusicSync.reconnect()
            call.resolve()
        }
    }
}

File: android/app/src/main/java/app/unimusic/UniMusicSyncPlugin.kt
package app.unimusic

import android.net.Uri
import android.os.Environment
import androidx.documentfile.provider.DocumentFile
import app.unimusic.sync.UniMusicSync
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancelChildren
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

@Suppress("unused")
@CapacitorPlugin(name = "UniMusicSync")
class UniMusicSyncPlugin : Plugin() {
    private val appJob = SupervisorJob()
    private val appScope = CoroutineScope(Dispatchers.IO + appJob)

    lateinit var uniMusicSync: UniMusicSync
        private set
    lateinit var uniMusicSyncPath: String


    override fun handleOnDestroy() {
        super.handleOnDestroy()

        appJob.cancelChildren()
        if (::uniMusicSync.isInitialized) {
            runBlocking {
                uniMusicSync.shutdown()
            }
            uniMusicSync.close()
        }
    }

    override fun load() {
        super.load()
        appScope.launch {
            val documentsPath = context.getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS)!!
            uniMusicSyncPath = documentsPath.resolve("UniMusicSync").path
            uniMusicSync = UniMusicSync.create(uniMusicSyncPath);
        }
    }

    private inline fun handleCallExceptions(call: PluginCall, block: () -> Unit) {
        try {
            block()
        } catch (e: Exception) {
            call.reject(e.message)
        }
    }

    @PluginMethod
    fun createNamespace(call: PluginCall) {
        if (!::uniMusicSync.isInitialized) {
            call.reject("UniMusicSync is not initialized")
            return
        }

        appScope.launch {
            handleCallExceptions(call) {
                val namespace = uniMusicSync.createNamespace()
                val result = JSObject()
                result.put("namespace", namespace)
                call.resolve(result)
            }
        }
    }

    @PluginMethod
    fun deleteNamespace(call: PluginCall) {
        if (!::uniMusicSync.isInitialized) {
            call.reject("UniMusicSync is not initialized")
            return
        }

        val namespace = call.getString("namespace") ?: run {
            call.reject("You must provide a 'namespace' option")
            return
        }

        appScope.launch {
            handleCallExceptions(call) {
                uniMusicSync.deleteNamespace(namespace)
                call.resolve()
            }
        }
    }

    @PluginMethod
    fun getAuthor(call: PluginCall) {
        if (!::uniMusicSync.isInitialized) {
            call.reject("UniMusicSync is not initialized")
            return
        }

        appScope.launch {
            handleCallExceptions(call) {
                val namespace = uniMusicSync.createNamespace()
                val result = JSObject()
                result.put("namespace", namespace)
                call.resolve(result)
            }
        }
    }

    @PluginMethod
    fun getNodeId(call: PluginCall) {
        if (!::uniMusicSync.isInitialized) {
            call.reject("UniMusicSync is not initialized")
            return
        }

        appScope.launch {
            handleCallExceptions(call) {
                val nodeId = uniMusicSync.getNodeId()
                val result = JSObject()
                result.put("nodeId", nodeId)
                call.resolve(result)
            }
        }
    }

    @PluginMethod
    fun getFiles(call: PluginCall) {
        if (!::uniMusicSync.isInitialized) {
            call.reject("UniMusicSync is not initialized")
            return
        }

        val namespace = call.getString("namespace") ?: run {
            call.reject("You must provide a 'namespace' option")
            return
        }

        appScope.launch {
            handleCallExceptions(call) {
                val files = uniMusicSync.getFiles(namespace)
                val result = JSObject()

                for (entry in files) {
                    if (entry.isEmpty()) {
                        continue
                    }

                    val fileObject = JSObject()
                    fileObject.put("key", entry.key())
                    fileObject.put("author", entry.author())
                    fileObject.put("timestamp", entry.timestamp())
                    fileObject.put("contentHash", entry.contentHash())
                    fileObject.put("contentLen", entry.contentLen())

                    result.put(entry.key(), fileObject)
                }

                call.resolve(result)
            }
        }
    }


    @PluginMethod
    fun writeFile(call: PluginCall) {
        if (!::uniMusicSync.isInitialized) {
            call.reject("UniMusicSync is not initialized")
            return
        }

        val namespace = call.getString("namespace") ?: run {
            call.reject("You must provide a 'namespace' option")
            return
        }

        val syncPath = call.getString("syncPath") ?: run {
            call.reject("You must provide a 'syncPath' option")
            return
        }

        // sourcePath in case of Android is a DocumentFile Uri
        val sourcePath = call.getString("sourcePath") ?: run {
            call.reject("You must provide a 'sourcePath' option")
            return
        }

        val fileUri = try {
            Uri.parse(sourcePath)
        } catch (e: Exception) {
            System.err.println(e)
            call.reject("'sourcePath' option has to be a valid SAF file uri")
            return
        }

        val file = DocumentFile.fromSingleUri(context, fileUri) ?: run {
            call.reject("Provided 'sourcePath' does not seem to exist")
            return
        }

        appScope.launch {
            handleCallExceptions(call) {
                val inputStream = context.contentResolver.openInputStream(file.uri) ?: run {
                    call.reject("Failed to open input stream for ${file.uri}")
                    return@launch
                }
                val data = inputStream.use { inputStream.readBytes() }
                val fileHash = uniMusicSync.writeFile(namespace, syncPath, data)
                val result = JSObject()
                result.put("fileHash", fileHash)
                call.resolve(result)
            }
        }
    }

    @PluginMethod
    fun deleteFile(call: PluginCall) {
        if (!::uniMusicSync.isInitialized) {
            call.reject("UniMusicSync is not initialized")
            return
        }

        val namespace = call.getString("namespace") ?: run {
            call.reject("You must provide a 'namespace' option")
            return
        }

        val syncPath = call.getString("syncPath") ?: run {
            call.reject("You must provide a 'syncPath' option")
            return
        }

        appScope.launch {
            handleCallExceptions(call) {
                uniMusicSync.deleteFile(namespace, syncPath)
                call.resolve()
            }
        }
    }

    @PluginMethod
    fun readFile(call: PluginCall) {
        if (!::uniMusicSync.isInitialized) {
            call.reject("UniMusicSync is not initialized")
            return
        }

        val namespace = call.getString("namespace") ?: run {
            call.reject("You must provide a 'namespace' option")
            return
        }

        val syncPath = call.getString("syncPath") ?: run {
            call.reject("You must provide a 'syncPath' option")
            return
        }

        appScope.launch {
            handleCallExceptions(call) {
                val data = uniMusicSync.readFile(namespace, syncPath)
                val result = JSObject()
                // TODO: Serve the file
                result.put("url", "")
                call.resolve(result)
            }
        }
    }

    @PluginMethod
    fun readFileHash(call: PluginCall) {
        if (!::uniMusicSync.isInitialized) {
            call.reject("UniMusicSync is not initialized")
            return
        }

        val fileHash = call.getString("fileHash") ?: run {
            call.reject("You must provide a 'fileHash' option")
            return
        }

        appScope.launch {
            handleCallExceptions(call) {
                val data = uniMusicSync.readFileHash(fileHash)
                val result = JSObject()
                // TODO: Serve the file
                result.put("url", "")
                call.resolve(result)
            }
        }
    }

    @PluginMethod
    fun export(call: PluginCall) {
        if (!::uniMusicSync.isInitialized) {
            call.reject("UniMusicSync is not initialized")
            return
        }

        val namespace = call.getString("namespace") ?: run {
            call.reject("You must provide a 'namespace' option")
            return
        }

        val syncPath = call.getString("syncPath") ?: run {
            call.reject("You must provide a 'syncPath' option")
            return
        }

        // destinationPath in case of Android is a DocumentFile Uri
        val destinationPath = call.getString("destinationPath") ?: run {
            call.reject("You must provide a 'destinationPath' option")
            return
        }

        val fileUri = try {
            Uri.parse(destinationPath)
        } catch (e: Exception) {
            System.err.println(e)
            call.reject("'destinationPath' option has to be a valid SAF file uri")
            return
        }

        val file = DocumentFile.fromSingleUri(context, fileUri) ?: run {
            call.reject("Provided 'destinationPath' does not seem to exist")
            return
        }

        appScope.launch {
            handleCallExceptions(call) {
                // Since Android<->Rust filesystem interop would be a pain in the ass,
                // we do similar things to what Rust does directly here
                val outputStream = context.contentResolver.openOutputStream(file.uri) ?: run {
                    call.reject("Failed to open input stream for ${file.uri}")
                    return@launch
                }
                val data = uniMusicSync.readFile(namespace, syncPath)
                outputStream.use { outputStream.write(data) }
                call.resolve()
            }
        }
    }

    @PluginMethod
    fun exportHash(call: PluginCall) {
        if (!::uniMusicSync.isInitialized) {
            call.reject("UniMusicSync is not initialized")
            return
        }

        val fileHash = call.getString("fileHash") ?: run {
            call.reject("You must provide a 'fileHash' option")
            return
        }

        // destinationPath in case of Android is a DocumentFile Uri
        val destinationPath = call.getString("destinationPath") ?: run {
            call.reject("You must provide a 'destinationPath' option")
            return
        }

        val fileUri = try {
            Uri.parse(destinationPath)
        } catch (e: Exception) {
            System.err.println(e)
            call.reject("'destinationPath' option has to be a valid SAF file uri")
            return
        }

        val file = DocumentFile.fromSingleUri(context, fileUri) ?: run {
            call.reject("Provided 'destinationPath' does not seem to exist")
            return
        }

        appScope.launch {
            handleCallExceptions(call) {
                // Since Android<->Rust filesystem interop would be a pain in the ass,
                // we do similar things to what Rust does directly here
                val outputStream = context.contentResolver.openOutputStream(file.uri) ?: run {
                    call.reject("Failed to open input stream for ${file.uri}")
                    return@launch
                }
                val data = uniMusicSync.readFileHash(fileHash)
                outputStream.use { outputStream.write(data) }
                call.resolve()
            }
        }
    }

    @PluginMethod
    fun share(call: PluginCall) {
        if (!::uniMusicSync.isInitialized) {
            call.reject("UniMusicSync is not initialized")
            return
        }

        val namespace = call.getString("namespace") ?: run {
            call.reject("You must provide a 'namespace' option")
            return
        }

        appScope.launch {
            handleCallExceptions(call) {
                val ticket = uniMusicSync.share(namespace)
                val result = JSObject()
                result.put("ticket", ticket)
                call.resolve(result)
            }
        }
    }

    @PluginMethod
    fun import(call: PluginCall) {
        if (!::uniMusicSync.isInitialized) {
            call.reject("UniMusicSync is not initialized")
            return
        }

        val ticket = call.getString("ticket") ?: run {
            call.reject("You must provide a 'ticket' option")
            return
        }

        appScope.launch {
            handleCallExceptions(call) {
                val namespace = try {
                    uniMusicSync.import(ticket)
                } catch (e: Exception) {
                    System.err.println(e)
                    call.reject("Failed to import ticket")
                    return@launch
                }

                val result = JSObject()
                result.put("namespace", namespace)
                call.resolve(result)
            }
        }
    }

    @PluginMethod
    fun sync(call: PluginCall) {
        if (!::uniMusicSync.isInitialized) {
            call.reject("UniMusicSync is not initialized")
            return
        }

        val namespace = call.getString("namespace") ?: run {
            call.reject("You must provide a 'namespace' option")
            return
        }

        appScope.launch {
            handleCallExceptions(call) {
                try {
                    uniMusicSync.sync(namespace)
                } catch (e: Exception) {
                    System.err.println(e)
                }
                call.resolve()
            }
        }
    }

    @PluginMethod
    fun reconnect(call: PluginCall) {
        if (!::uniMusicSync.isInitialized) {
            call.reject("UniMusicSync is not initialized")
            return
        }

        appScope.launch {
            handleCallExceptions(call) {
                uniMusicSync.reconnect()
                call.resolve()
            }
        }
    }
}